src/Microsoft.ML.FastTree/GamTrainer.cs

```diff
@@ -240,45 +143,71 @@ private protected GamTrainerBase(IHostEnvironment env, TArgs args)
             Host.CheckParam(0 <= args.GainConfidenceLevel && args.GainConfidenceLevel < 1, nameof(args.GainConfidenceLevel), "Must be in [0, 1).");
             Host.CheckParam(0 < args.MaxBins, nameof(args.MaxBins), "Must be posittive.");
             Host.CheckParam(0 < args.NumIterations, nameof(args.NumIterations), "Must be positive.");
+            Host.CheckParam(0 < args.MinDocuments, nameof(args.MinDocuments), "Must be positive.");
 
             Args = args;
-            Info = new TrainerInfo(normalization: false, calibration: NeedCalibration, caching: false);
+            Info = new TrainerInfo(normalization: false, calibration: NeedCalibration, caching: false, supportValid: true);
             _gainConfidenceInSquaredStandardDeviations = Math.Pow(ProbabilityFunctions.Probit(1 - (1 - Args.GainConfidenceLevel) * 0.5), 2);
             _entropyCoefficient = Args.EntropyCoefficient * 1e-6;
+
+            ParallelTraining = new SingleTrainer();
+
             int numThreads = args.NumThreads ?? Environment.ProcessorCount;
             if (Host.ConcurrencyFactor > 0 && numThreads > Host.ConcurrencyFactor)
-            {
                 using (var ch = Host.Start("GamTrainer"))
                 {
                     numThreads = Host.ConcurrencyFactor;
                     ch.Warning("The number of threads specified in trainer arguments is larger than the concurrency factor "
                         + "setting of the environment. Using {0} training threads instead.", numThreads);
                     ch.Done();
                 }
-            }
 
             InitializeThreads(numThreads);
         }
 
-        public sealed override TPredictor Train(TrainContext context)
+        /// <summary>
+        /// Load the data off the context and train the model
+        /// </summary>
+        /// <param name="context"></param>
+        protected void TrainBase(TrainContext context)
         {
             using (var ch = Host.Start("Training"))
             {
                 ch.CheckValue(context, nameof(context));
-                ConvertData(context.TrainingSet);
+
+                // Create the datasets
+                ConvertData(context.TrainingSet, context.ValidationSet);
+
+                // Define scoring and testing
+                DefineScoreTrackers();
+                if (HasValidSet)
+                    DefinePruningTest();
                 InputLength = context.TrainingSet.Schema.Feature.Type.ValueCount;
+
                 TrainCore(ch);
-                var pred = CreatePredictor();
+
                 ch.Done();
-                return pred;
             }
         }
 
-        private protected abstract TPredictor CreatePredictor();
+        /// <summary>
+        /// Define the ScoreTrackers to use in the calculation.
```

Comment:
>Define the ScoreTrackers [](start = 12, length = 24)

these comments are not helpful. I already have a good suspicion that `DefineScoreTrackers` defines score trackers. 

These are private methods too, so might as well not have comments at all

Reply:
Thanks -- removed!